File: herald-admin-ruby/lib/herald_admin/rooms.rb
# frozen_string_literal: true

module HeraldAdmin
  class RoomNamespace
    def initialize(transport)
      @t = transport
    end

    def create(id, name, meta: nil)
      body = { id: id, name: name }
      body[:meta] = meta if meta
      data = @t.request("POST", "/rooms", body)
      Room.new(**data.transform_keys(&:to_sym))
    end

    def get(id)
      data = @t.request("GET", "/rooms/#{ERB::Util.url_encode(id)}")
      Room.new(**data.transform_keys(&:to_sym))
    end

    def update(id, name: nil, meta: nil)
      body = {}
      body[:name] = name if name
      body[:meta] = meta if meta
      @t.request("PATCH", "/rooms/#{ERB::Util.url_encode(id)}", body)
    end

    def list
      data = @t.request("GET", "/rooms")
      data["rooms"].map { |r| Room.new(**r.transform_keys(&:to_sym)) }
    end

    def delete(id)
      @t.request("DELETE", "/rooms/#{ERB::Util.url_encode(id)}")
    end
  end
end

File: herald-admin-typescript/src/namespaces/rooms.ts
import type { HttpTransport } from "../transport.js";
import type { Stream } from "../types.js";

export class StreamNamespace {
  constructor(private transport: HttpTransport) {}

  async list(): Promise<Stream[]> {
    const res = await this.transport.request<{ streams: Stream[] }>("GET", "/streams");
    return res.streams;
  }

  async create(
    id: string,
    name: string,
    options?: { meta?: unknown; public?: boolean },
  ): Promise<Stream> {
    return this.transport.request<Stream>("POST", "/streams", {
      id,
      name,
      meta: options?.meta,
      public: options?.public,
    });
  }

  async get(id: string): Promise<Stream> {
    return this.transport.request<Stream>("GET", `/streams/${encodeURIComponent(id)}`);
  }

  async update(id: string, options: { name?: string; meta?: unknown; archived?: boolean }): Promise<void> {
    return this.transport.request("PATCH", `/streams/${encodeURIComponent(id)}`, options);
  }

  async delete(id: string): Promise<void> {
    return this.transport.request("DELETE", `/streams/${encodeURIComponent(id)}`);
  }
}
